src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -423,6 +427,70 @@ data CodingFailureMode
 replacementChar :: Char
 replacementChar = '\xFFFD'
 
+data UTF8CharDecodeState a
+    = UTF8CharDecodeInit
+    | UTF8CharDecoding !DecodeState !CodePoint
+
+-- XXX We can chain this with another parser to propagate the error and act
+-- accordingly.
+{-# INLINE parseCharUtf8D' #-}
+parseCharUtf8D' :: Monad m => ParserD.Parser m Word8 Char
+parseCharUtf8D' =
+    let A.Array _ ptr _ = utf8d
+    in ParserD.Parser (step' ptr) initial extract
+
+    where
+
+    {-# INLINE prefix #-}
+    prefix = "Streamly.Internal.Data.Stream.parseCharUtf8WithD:"
+
+    {-# INLINE initial #-}
+    initial = return $ ParserD.IPartial UTF8CharDecodeInit
+
+    {-# INLINE extract #-}
+    extract _ = error $ prefix ++ "Not enough input"
+
+    {-# INLINE step' #-}
+    step' table UTF8CharDecodeInit x =
+        -- Note: It is important to use a ">" instead of a "<=" test
+        -- here for GHC to generate code layout for default branch
+        -- prediction for the common case. This is fragile and might
+        -- change with the compiler versions, we need a more reliable
+        -- "likely" primitive to control branch predication.
+        case x > 0x7f of
+            False -> return $ ParserD.Done 0 $ unsafeChr $ fromIntegral x
+            True ->
+                let (Tuple' sv cp) = decode0 table x
+                 in case sv of
+                        12 ->
+                            let msg = prefix
+                                    ++ "Invalid first UTF8 byte" ++ show x
+                             in return $ ParserD.Error msg
+                        0 -> error $ prefix ++ "unreachable state"
+                        _ ->
+                            return $ ParserD.Continue 0 (UTF8CharDecoding sv cp)
+
+    step' table (UTF8CharDecoding statePtr codepointPtr) x =
+        let (Tuple' sv cp) = decode1 table statePtr codepointPtr x
+         in case sv of
+            0 -> return $ ParserD.Done 0 $ unsafeChr cp
+            12 ->
+                let msg = prefix
+                        ++ "Invalid subsequent UTF8 byte"
+                        ++ show x
+                        ++ "in state"
+                        ++ show statePtr
+                        ++ "accumulated value"
+                        ++ show codepointPtr
+                 in return $ ParserD.Error msg
+            _ -> return $ ParserD.Continue 0 (UTF8CharDecoding sv cp)
```

Comment:
return can be factored out before `case`.

In fact, since initial, step, extract all are pure we can use our new `parser` smart constructor to make this parser.

Reply:
I've had to introduce another constructor in the state for this. But that shouldn't be a problem I guess.